include/lsst/afw/table/python/catalog.h

```diff
@@ -144,17 +157,22 @@ void declareCatalog(pybind11::class_<CatalogT<RecordT>, std::shared_ptr<CatalogT
     cls.def("_extend", [](Catalog & self, Catalog const & other, SchemaMapper const & mapper) {
         self.insert(mapper, self.end(), other.begin(), other.end());
     });
-    cls.def("_append", [](Catalog & self, std::shared_ptr<RecordT> const & rec) {
+    cls.def("_append", [](Catalog & self, std::shared_ptr<Record> const & rec) {
         self.push_back(rec);
     });
     cls.def("_delitem_", [](Catalog & self, std::ptrdiff_t i) {
         self.erase(self.begin() + utils::python::cppIndex(self.size(), i));
     });
     cls.def("_clear", &Catalog::clear);
- 
+
     cls.def("set", &Catalog::set);
     cls.def("_getitem_", [](Catalog & self, int i) {
-        return self.get(utils::python::cppIndex(self.size(), i));
+        try { // try/catch is a workaround for DM-9715
+            return self.get(utils::python::cppIndex(self.size(), i));
+        } catch (pex::exceptions::OutOfRangeError & err) {
+            PyErr_SetString(PyExc_IndexError, err.what());
+            throw py::error_already_set();
```

Comment:
IMO, this is definitely better, but I may be a weird bird.  In any case, I expect it to be removed in the future.